### DISCUSSION TITLE
Decoding and Paging

### DISCUSSION THREAD

**Discussion by KenWillmott** [KenWillmott/HD6303EVB#9] [General]
I'm starting another thread to focus on memory mapping. With the 512x8 RAM (or more), there are great possibilities that it would a shame to waste. In keeping with the desire for low parts count, it would be good if it were simple too. It's a blessing that the 6303 has internal RAM, a small core function there can run independently of any external mapping. Code can not be running in memory that is being paged. That means the entire external memory space can be paged, enabled or disabled at will. An 8 bit latch can forward the 3 additional address bits needed for the 512k, not only that but the 5 extra bits can be used to address more memory, or for other memory mapping purposes. If it made resettable at boot time, a default boot configuration can be enforced.

Since the entire RAM is battery backed, the main purpose for EEPROM could be just to boot up, copy some code into some RAM areas, and then disable itself. This is known as "shadow ROM", and has the advantage that it makes the entire memory space (not including memory mapped peripherals) available as either code or data space. It could be very easy to implement, for example using one (the highest)? bit in the address latch. Or, the EEPROM could remain selected from boot time onwards if that feature is not desired. That would be a software option, if selected, the lower 32k of each of the 8 available memory pages would be selectable via the latch. Paging would still be possible with that, it's more versatile in that you can actually run code directly from paged memory, which you can't do using a RAM disk model. But that is all in the realm of software. The hardware idea is to try to present an effective interface that could be used either for RAM disk, or essentially virtual memory (except requiring to be implemented in software rather than hardware).

RAM disk supports the existing body of software for this processor, perhaps split in some users preference like 128k RAM, 384k RAM disk. Paged program memory when provided is normally done with an MMU and has software support. However I believe that some modular code might find a way to run in such a system.

All this would avoid the need for pages smaller than 64k... those would not be a problem per se, but I found that decoding for that can get complicated. Pages smaller than 64k require a mux to select between the paged and non-paged memory segments. 74xx157 I think.

In any case, the decode circuit only has to map an I/O space and its internal divisions, and produce a memory deselect signal. Generally, I'd like to see the I/O located as high in memory as possible. Sometimes it's placed "below" the EEPROM but that creates a RAM size limit, I'd like to avoid. With the shadow ROM (or really, any ROM), the I/O could be overlaid in the ROM space. It is only necessary to support the CPU vector table from $FFF8 up, at boot time. Everything else is flexible.

Another possibility, is to provide a jumper that enables the EEPROM, the firmware would copy itself into RAM one time only. Then, with boot firmware in NVRAM, there is no real reason for EEPROM except for crash recovery. Even that has a solution, it is possible to write protect areas of NVRAM although it demands more logic ICs.

It's a matter of finding a sweet spot between cool functionality and a minimum glue logic requirement.

**Comment by TheEccentricGenius** [comment]
My usage is just to use the EEPROM for extended times with the board off.

**Comment by KenWillmott** [comment]
I found another avenue that needs some serious consideration. Paging has been an important issue since the beginnings of the project. Because, with the fairly low cost and high density of static RAM now, it is a shame to not take full advantage. There are some things that we don't have any more, even if we hated them, we needed them, witness floppy drives. I wanted to use a minimum of parts and still use all of a large, e.g. 512k x 8 RAM efficiently. So I returned to the problem with only pen and paper, as my designs only seemed to be getting more and more complex.

What I found out, the absolutely easiest paging is done, not with many glue logic circuits as in previous concepts, but with two separate memory IC's. The EVB 512k upgrade does that, but I realized I can do one better. By using two 512k IC's, mapped to upper and lower memory respectively, a very simple paging mechanism can be supported, and yield a total 1M memory complement as a bonus. Looking deeper, I realized that it is likely possible to battery back up both of them, using only one NVRAM controller IC. 1Mb is roughly close to having two floppy discs so it mirrors the storage of legacy systems somewhat.

**Reply by TheEccentricGenius** [reply]
I love floppy discs....My custom OS for PCs fits into it easily.

I would like to see the new design....Just to keep the HD6303RP board aligned to it.

**Comment by KenWillmott** [comment]
Current decoding scheme is this:
![image](https://github.com/KenWillmott/HD6303EVB/assets/17345651/f67fb803-a824-4f9a-ac9e-ac69a747bc79)
